Feature/improved imputer

icedoom888 opened the pull request:
Fixes #73

JesperDramsch commented on src/anemoi/models/preprocessing/imputer.py:
```diff
@@ -109,11 +112,8 @@ def transform(self, x: torch.Tensor, in_place: bool = True) -> torch.Tensor:
-            idx = [slice(0, 1)] * (x.ndim - 2) + [slice(None), slice(None)]
-            self.nan_locations = torch.isnan(x[idx].squeeze())
+        # Initilialize mask every time
+        self.nan_locations = torch.isnan(x)
```
This change likely breaks for different batch sizes in training, validation, and inference. What is the reasoning behind removing the slicing?

JesperDramsch commented on src/anemoi/models/models/hierarchical.py:
Should this be introduced in this PR?

JesperDramsch commented on src/anemoi/models/preprocessing/imputer.py:
```diff
@@ -129,11 +129,22 @@ def transform(self, x: torch.Tensor, in_place: bool = True) -> torch.Tensor:
+
+        return x
+    
+        # TODO: Does this inverse transform make sense? Why are we inserting NaNs again? what would we do at inference time? 
```
Philosophically, the model should "give back what it got".

JesperDramsch reviewed (commented):
Hi @icedoom888, thanks for taking on the idea of the flexible imputer.

## Philosophical Considerations
I have had some discussions with internal colleagues around the idea of a fully time-variant imputer and they've opened my eyes to even more issues with the idea. Essentially, the model should "return what it got" so when you feed NaNs it should also return NaNs in the right locations. However, for time-varying NaN locations, this would mean that during inference, we need another model that predicts the NaN locations to fill (or some process to fill those), as there is no automated way to know where a NaN location would be during inference. 

With that being said, comments regarding the code.

## Code comments
1. It looks like you also checked in the hierarchical processor, is that intended?
2. You made modifications to the code that are unfortunately wrong and will break at least two things: Varying batch sizes and flexible Preprocessors.

That unfortunately means we cannot implement these changes as they stand. I have left inline comments where these assumptions are made and how they break the flexibility of the code and affect the existing `ConstantImputer` and `InputImputer`.

Maybe it would be better to discuss the viability of a Flexible Imputer first in #73 before we restart this implementation?

icedoom888 commented:
Hey @JesperDramsch! Thanks for reviewing this ;) 

I fixed both breaking issues you mentioned in your review! Code should be fine now.

Concerning this: 
I believe the imputer should be used as a fix for NaNs in the data, not as a predictior. 
The model itself will learn how to read and predict imputed values if they are not masked out in the loss.

If we train over imputed values, this will intrinsecally solve the issue to predict NaNs at inference time.

icedoom888 commented:
@JesperDramsch I added all changed we discussed and tested everything works as expected 💯
I added the warning as we discussed but feel free to change it up to whatever you think it's the most informative!

JesperDramsch reviewed (approved):
Thanks for adding implementing all the changes. It looks really nice now
